lib/transformer.js

```diff
@@ -0,0 +1,288 @@
+    self.applyRule(rule, ruleCallback);
+  }, function (err) {
+    cb(err, self);
+  });
+};
+
+/**
+ * Adds a rule generated warning to the transformer.
+ * @param object Object that generated the warning.
+ * @param msg Message for the warning.
+ */
+Transformer.prototype.addWarning = function (object, msg) {
+  this.warnings.push(new Warning(object, msg));
+};
+
+/**
+ * Allows the user to override existing rule actions and define new ones.
+ * Note the action function will be applied within the context of this
+ * Transformation.
+ * @param {string} name Name of the action to define or override.
+ * @param {fs-transform~Rule} fn Function to apply when encountering the rule
+ *   with the given action name.
+ */
+Transformer.prototype.setAction = function(name, fn) {
+  var self = this;
+  this._ruleActions[name] = function() {
+    fn.apply(self, arguments);
+  };
+};
+
+/**
+ * Determine the rule action handler for the action name.
+ * @param {string} name Name of the action.
+ * @return {function} The rule handler for the action.
+ */
+Transformer.prototype.getAction = function(name) {
+  return this._ruleActions[name];
+};
+
+/**
+ * Callback for the application of rule actions.
+ * @callback fs-transform~Rule
+ * @param {object} rule Rule to apply.
+ * @param {function} cb Callback to execute after the rule has been applied.
+ */
+
+/**
+ * Applys a given transformation rule.
+ * @param {object} rule Rule to apply.
+ * @param {function} cb Callback to execute once the rule has been processed.
+ */
+Transformer.prototype.applyRule = function (rule, cb) {
+  if (!rule) {
+    return cb(new Error('Null or undfined rule encountered.'));
+  }
+
+  if (!isString(rule.action)) {
+    return cb(new Error('Invalid rule action: ' + rule.action));
+  }
+
+  var actionMethod = this.getAction(rule.action);
+  if (!actionMethod) {
+    return cb(new Error('Action not found: ' + rule.action));
+  }
+
+  actionMethod(rule, cb);
+};
+
+/**
+ * Copy a file.
+ * @param {object} rule Defines the source and destination for the file copy.
+ * @param {function} cb Callback to execute once the file has been copied.
+ */
+Transformer.prototype.copy = function (rule, cb) {
+  if (!isString(rule.source)) {
+    this.addWarning(rule, 'Missing source file.');
+    return cb();
+  }
+  if (!isString(rule.dest)) {
+    this.addWarning(rule, 'Missing destination file.');
+    return cb();
+  }
+  if (!this.driver.exists(rule.source)) {
+    this.addWarning(rule, 'Source file does not exist.');
+    return cb();
+  }
+  if (this.driver.exists(rule.dest)) {
+    this.addWarning(rule, 'Overwriting destination file.');
+  }
+  this.driver.copy(rule.source, rule.dest, cb);
+};
+
+/**
+ * Rename a file.
+ * @param {object} rule Defines the source and destination for the file rename.
+ * @param {function} cb Callback to execute once the file has been renamed.
+ */
+Transformer.prototype.rename = function (rule, cb) {
```

Comment:
Yeah I was thinking about making a 'binary command' method or something like that. But I was thinking that while it's not very DRY to do it this way, it is definitely more clear to anyone reading the source.